DM-31295: Cache boto client and add timer messages to logs for some file I/O

timj opened the pull request:
## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

timj committed:
Add context manager to time some code

timj committed:
Fix log messages that were not delaying string interpolation

timj committed:
Add timing information to logs for file download/upload/read

timj committed:
Cache the s3 client

This makes a huge difference.

ktlim commented on python/lsst/daf/butler/core/_butlerUri/file.py:
```diff
@@ -217,8 +217,11 @@ def transfer_from(self, src: ButlerURI, transfer: str,
 
-        log.debug(f"Transferring {src} [exists: {src.exists()}] -> "
-                  f"{self} [exists: {self.exists()}] (transfer={transfer})")
+        # Existence checks can take time so only try if the log message
```
In C++, log macros ensure that log message components are not even evaluated until after it is decided that the message will be issued.  There's no equivalent in Python?  Maybe the log method could take a callable that returns a dict of message components and that would only be called after the threshold check.

timj replied:
I've not seen anyone doing deferred calling of log message components in python. We could pass `self.exists` as an argument to some object which then calls it later. This is a bit like how the python docs suggest f-string formatting of log messages is supported: https://docs.python.org/3/howto/logging-cookbook.html#using-custom-message-objects -- can I not worry about this for today and leave it with the isEnabledFor?

ktlim replied:
Of course.  I was just complaining about the state of Python logging, not asking you to fix it :)

ktlim commented on python/lsst/daf/butler/core/_butlerUri/http.py:
```diff
@@ -340,9 +343,10 @@ def read(self, size: int = -1) -> bytes:
             The number of bytes to read. Negative or omitted indicates
             that all data should be read.
         """
-        log.debug("Reading from remote resource: %s", self.geturl())
```
You don't think it could still be useful to know what the URL is, separate from timing?

timj replied:
I was vacillating on this. I know that the timer message will come out regardless so I thought that seeing log output of:
```
DEBUG reading URI X
DEBUG read URI X took Y seconds
```
was redundant. Sounds like you'd prefer the log message showing the intent as well as the log message with the result.

timj replied:
I suppose one argument in favor of the duplication is that they are in fact different loggers so can be configured with different levels.

ktlim replied:
> different loggers

Exactly.

ktlim commented on python/lsst/daf/butler/core/utils.py:
```diff
@@ -475,3 +478,58 @@ def isplit(string: T, sep: T) -> Iterator[T]:
+        Python logging level to use to issue the log message. If the
+        code block raises an exception the log message will automatically
+        switch to level ERROR.
+    log_prefix : `str`, optional
```
Your invocations are verbose enough that I wonder if you want to simplify by defaulting to prefixing with `timer`.

timj replied:
Okay. I'll default it to `timer`. I was also wondering about dropping the `log_` from the variable names so maybe I should do that as well.

ktlim commented:
Maybe change the PR title to reflect the addition of session caching?

timj committed:
Default time_this to use prefix="timer" and drop log_ name in params

We want to always use "timer" as the prefix so set it as the
default. Also change log_prefix -> prefix and log_level -> level
